_labs/lab08.markdown

```diff
@@ -0,0 +1,171 @@
+- `alarm(inst)` : There exists a potential exploitable divide-by-zero error at instruction inst.
+
+
+Assume that input programs may contain function calls to `tainted_input` and `sanitizer` that read and sanitize a tainted input, respectively. The final output relation for potential bug reports is `alarm`.
+
+You will use these relations to build rules for the definition analysis and taint analysis in `Extractor.cpp`.
+
+**Encoding LLVM Instruction in Datalog**. Recall that, in LLVM IR, values and instructions are interchangable. Therefore, all variables X, Y, and Z are an instance of LLVM’s `Value` class. 
+
+Assume that input C programs do not have pointer variables. Therefore, we abuse pointer variables in LLVM IR as their dereference expressions. Consider the following simplified LLVM program from a simple C program `int x = 0; int y = x;`:
+
+```sh
+x = alloca i32          ; I0
+y = alloca i32          ; I1
+store i32 0, i32* x     ; I2
+a = load i32, i32* x    ; I3
+store i32 a, i32* y     ; I4
+```
+We ignore alloca instructions and consider that each store instruction defines the second argument.
+In the case of the above example, you should have `Def(I0,I2)`, because x corresponds to x in LLVM IR. Likewise, consider each load instruction uses the argument.
+In the example, you should have `Use(I0,I3)` and `Def(I3,I3)` because load instructions define a non-pointer variable which is represented as the instruction itself in LLVM.
+Finally, you should have `Use(I3,I4)` and `Def(I1,I4)` for instruction I4.
+
+**Defining Datalog Rules from C++ API**. You will write your Datalog rules in the function `initialize` using the relations above. Consider  an example Datalog rule:
+
+A(X, Y) :- B(X, Z), C(Z, Y).
+
+This rule corresponds to the following formula:
+X,Y,Z. B(X,Z) C(Z,Y) A(X,Y).
+
+In Z3, you can specify the formula in the following sequence of APIs in `initialize`:
+
+```sh
+/* Declare quantified variables */
+z3::expr X = C.bv_const(“X”, 32);   // encode X as a 32-bit bitvector (bv)
+z3::expr Y = C.bv_const(“Y”, 32);
+z3::expr Z = C.bv_const(“Z”, 32);
+/* Define and register rules */
+z3::expr R0 = z3::forall(X, Y, Z, z3::implies(B(X,Z) && C(Z, Y), A(X,Y)));
+Solver->add_rule(R0, C.str_symbol(“R0”));
+```
+
+You might take a look at the important classes including [expr][expression] and [fixed point][fixed-point] as well as [an example][example] of using Z3 C++ API.
+
+**Extracting Datalog Facts**. You will need to implement the function `extractConstraints` in 'Extractor.cpp' to extract Datalog facts for each LLVM instruction. The skeleton code provides a couple of auxiliary functions in `lab8/src/Extract.cpp` and `lab8/src/Utils.cpp` help you with this task:
+
+- `void addX(const InstMapTy &InstMap, ...)`
+- - X denotes the name of a relation. These functions add a fact of X to the solver. It takes `InstMap` that encodes each LLVM instruction as an integer. This map is initialized in the `main` function.
```

Comment:
These functions add a fact of X to the solver -> These functions add a fact of X to the facts file.